xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/LocalMUCUser.java

```diff
@@ -574,6 +575,71 @@ private void processDecliningInvitationMessage(
                                          info.elementTextTrim("reason"), packet.getFrom());
     }
 
+    /**
+    * checks for memberlist retrieval query
+    * @return true if occupant is allowed to retrieve member list otherwise false
+    */
+    private boolean checkForMemberListQuery(IQ packet)
+    {
+        MUCRoom room = server.hasChatRoom(packet.getTo().getNode())?server.getChatRoom(packet.getTo().getNode()):null;
+        if (room!=null&&(room.getOwners().contains(packet.getFrom().asBareJID())||
+            room.getAdmins().contains(packet.getFrom().asBareJID())||
+            room.getMembers().contains(packet.getFrom().asBareJID()))||
+            server.getSysadmins().contains(packet.getFrom().asBareJID())||
+            XMPPServer.getInstance().getAdmins().contains(packet.getFrom().asBareJID()))
+        {
+            Element query = packet.getChildElement();
+            if (query!=null&&query.getName().equalsIgnoreCase("query"))
```

Comment:
Why is this check necessary? This seems to indicate that "permission" to do something is based on what the client sends. That does not sound safe to me, as we can't verify that what a client sends is truthful (it can easily be abused).

Reply:
ok I've seperated the two checks.. it now checks for the listretrieval and then it will check if non-occupant is allowed to retrieve the list